[analyzer] Remove redundant bug type DoubleDelete

NagyDonat opened the pull request:
This commit removes the DoubleDelete bug type from `MallocChecker.cpp` because it's completely redundant with the `DoubleFree` bug (which is already used for all allocator families, including new/delete).

This simplifies the code of the checker and prevents the potential confusion caused by two semantically equivalent and very similar, but not identical bug report messages.

NagyDonat committed:
[analyzer] Remove redundant bug type DoubleDelete

This commit removes the DoubleDelete bug type from `MallocChecker.cpp`
because it's completely redundant with the `DoubleFree` bug (which is
already used for all allocator families, including new/delete).

This simplifies the code of the checker and prevents the potential
confusion caused by two semantically equivalent and very similar, but
not identical bug report messages.

NagyDonat committed:
Satisfy git-clang-format

steakhal commented on clang/lib/StaticAnalyzer/Checkers/MallocChecker.cpp:
```diff
@@ -3324,7 +3298,7 @@ void MallocChecker::checkUseZeroAllocated(SymbolRef Sym, CheckerContext &C,
 
   if (isReleased(Sym, C)) {
-    HandleDoubleDelete(C, Sym);
+    HandleDoubleFree(C, SourceRange(), /*Released=*/true, Sym, nullptr);
```
Can you use named param passing for the `nullptr`?

NagyDonat replied:
Done in https://github.com/llvm/llvm-project/pull/147542/commits/ffbafe4aeabe794c198169a53c67bf2ab60057ea

steakhal commented on clang/test/Analysis/NewDelete-checker-test.cpp:
```diff
@@ -412,7 +412,7 @@ class DerefClass{
   DerefClass *foo = new DerefClass();
   delete foo;
-  delete foo; // newdelete-warning {{Attempt to delete released memory}}
+  delete foo; // newdelete-warning {{Attempt to free released memory}}
```
I'm not convinced that the message change is an improvement.
Can we somehow still use the `delete` spelling?

steakhal replied:
One idea is to inspect the Expr against which we report the bug is a DeleteExpr or a callexpr.

NagyDonat replied:
I see your point and agree that _in that particular case_ the spelling with `free` is slightly worse than the old message that said `delete`, but I think that overall the _consistently_ slightly worse message is still better than the inconsistent situation where the checker could emit different messages in almost-identical situations.

I'll try to ensure that in the foreseeable future _somebody_ (e.g. an intern in our team) systemically updates the messages to use properly generic vocabulary and e.g. say "release" instead of "free".

NagyDonat committed:
Use named param passing for PrevSym

NagyDonat committed:
Inline checkDoubleDelete, properly document the role of this branch

I decided to inline and remove `checkDoubleDelete` because (1) it was
fairly short and (2) it is not a natural "do what is reasonable in
<a certain situation>" method so we cannot hide its implementation
behind a clear and descriptive name.
